### ISSUE TITLE
Quantized Conv2D op gives different result in TensorFlow and TFLite

### ISSUE THREAD

**Issue by ChanZou** [tensorflow/tensorflow#38845]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
`No`
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
`Linux Ubuntu 18.04`
- TensorFlow installed from (source or binary):
`pip`
- TensorFlow version (use command below):
`2.0.0`
- Python version:
`3.7.7`
- GPU model and memory:
`CPU only`

**Describe the current behavior**
Conv2D op from the quanitzation-aware training graph yields different results when converted to TensorFlow Lite and executed in uint8. Manual quantization has been done when comparing the float value to the uint8 value. Both TF and TFLite models are run on x86-64 CPU.
**Describe the expected behavior**
The op should give exactly the same result when comparing the node output of the TFLite node to the output of the fake-quantization node in the TF graph to ensure the correctness of the TFLite model.

**Standalone code to reproduce the issue**
```import os
import numpy as np
import tensorflow as tf
from imageio import imread
from typing import Any, AnyStr, Mapping, Sequence


def saved_model_to_tflite(saved_model_dir: AnyStr, output_path: AnyStr):
    converter = tf.compat.v1.lite.TFLiteConverter.from_saved_model(
        saved_model_dir=saved_model_dir)
    converter.inference_type = tf.uint8
    converter.inference_input_type = tf.uint8

    converter.quantized_input_stats = {"ToFloat": [127, 127.5]}
    converter.optimizations = []
    converter.change_concat_input_ranges = False
    converter.reorder_across_fake_quant = False
    tflite_model = converter.convert()

    with open(output_path, 'wb') as f:
        f.write(tflite_model)


def compare_models(saved_model_dir: AnyStr, tflite_path: AnyStr, image_path: AnyStr):
    image = imread(image_path)
    image = np.ascontiguousarray(image, dtype=np.float32)
    image = image[np.newaxis, ...]
    normalized_image = ((image - 127) / 127.5)
    
    imported_model = tf.saved_model.load(saved_model_dir)
    imported_model.variables.trainable = False
    model = imported_model.signatures[tf.compat.v1.saved_model.signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY]
    
    interpreter = tf.lite.Interpreter(tflite_path)
    interpreter.allocate_tensors()
    interpreter.tensor(interpreter.get_input_details()[0]["index"])()[:, :, :, :] = image.astype(np.uint8)
    
    tf_result = list(model(tf.constant(normalized_image)).values())[0].numpy()
    interpreter.invoke()
    tflite_result = interpreter.tensor(interpreter.get_output_details()[0]["index"])().copy()
    # Manually apply de-quantization
    tflite_result = (tflite_result.astype(np.float) - 107) * 0.1115700826048851
 
    ae = np.abs(tf_result - tflite_result)
    max_ae = ae.max()
    max_diff_indices = np.argwhere(ae == np.max(ae))
    print(max_diff_indices)
    print(f"Max absolute error: {max_ae}. Example values tf: {tf_result[ae==np.max(ae)][0]}; tflite: {tflite_result[ae==np.max(ae)][0]}")
```
**Other info / logs** Include any logs or source code that would be helpful to
TensorFlow and TFLite model visualization of the minimum model containing only one Conv2d op:
![Screenshot from 2020-04-23 14-37-54](https://user-images.githubusercontent.com/10414613/80136811-3af22300-8570-11ea-8eb5-e936e8e22251.png)
![Screenshot from 2020-04-23 14-31-10](https://user-images.githubusercontent.com/10414613/80136822-3ded1380-8570-11ea-8ef7-f1eee7211c0d.png)

A minimum example including a saved model a converted TFLite model and an image has been attached to help reproduce the issue.
[tf_issue.zip](https://github.com/tensorflow/tensorflow/files/4524528/tf_issue.zip)

**Comment by thaink** [comment]
@ChanZou I think you are doing the opposite way by feeding the image to tflite and normalized_image to TF. Could you tried it again? And provide what is differences and max error between them?

**Comment by ChanZou** [comment]
@thaink Thank you for your reply. I have run the requested test and here are the results:
```
TF normalized; TFLite not normalized: Max absolute error: 0.11157090216875076. Example values tf: 4.685942649841309; tflite: 4.797513552010059; Number of 'max difference' within 2e-6 range: 225
TF not normalized; TFLite normalized: Max absolute error: 24.880128353834152. Example values tf: -11.93799877166748; tflite: 12.942129582166672; Number of 'max difference' within 2e-6 range: 149
```
And two more just in case:
```
TF normalized; TFlite normalized: Max absolute error: 15.396671414375305. Example values tf: -10.04130744934082; tflite: 5.355363965034485; Number of 'max difference' within 2e-6 range: 1
TF not normalized; TFLite not normalized: Max absolute error: 20.86360428482294. Example values tf: 16.512371063232422; tflite: -4.351233221590519; Number of 'max difference' within 2e-6 range: 1
```
I might be misunderstanding it but I thought what `quantized_input_stats` does is feeding normalization information into the input node of the TFLite model so that raw image value (uint8 in TFLite model) can be correctly mapped to normalized input value (float in TF model)

**Comment by suharshs** [comment]
Your first usage here is correct: 

    TF normalized; TFLite not normalized: Max absolute error: 0.11157090216875076. Example values tf: 4.685942649841309; tflite: 4.797513552010059; Number of 'max difference' within 2e-6 range: 225

FakeQuant may have some difference if the model is not trained sufficiently long to allow the weights to learn about the quantization error.

**Comment by ChanZou** [comment]
Here I am trying to compare Conv2D op in TF and TFLite given identical input and weights. In other words, if we are comparing 1. original TF model 2. fully quantization-aware trained TF model 3. converted TFLite model for #1 and 4. converted TFLite model for #2, we should expect differences between #1 and #2 (#3 and #4 as well) but not between #1 and #3 (or #2 and #4), which is what I am showing in this example. Again, thank you for your reply!

**Comment by suharshs** [comment]
I see, thanks for the explanation.

It seems we need to trace down if the quantization error is coming from the weight, or the output quantization, or when tflite quantizes the bias to int32 during conversion, to see if this is fundamental to the implementation of rescale, or a bug in how rescale is being performed.

We will take a look and update the issue, thanks!

**Comment by liufengdb** [comment]
Hi ChanZou, is the virtualization actually reflects your model? I couldn't find the FakeQuant op for the weight. 

On the other hand, the max abs output error is `0.11157090216875076`, which is very close to the scale value of the output tensor (0.1115700826048851). So it is possible due to the some rounding errors.

**Comment by ChanZou** [comment]
Hi liufengdb, my bad not expanding scopes exhaustively, here is the updated one. As shown in the graph both weights and the activate are quantized. Do I need to attach a fakquant node to the bias as well? While it might be reasonable to do so, the converter did not do that for me.
![Screenshot from 2020-05-04 17-04-12](https://user-images.githubusercontent.com/10414613/81014706-5e985180-8e2b-11ea-9194-f40e8f4af46a.png)

And I agree that possibly deep down this might be a rounding error, but as NNs nowadays tend to be very deep, perturbation like this will very likely lead to catastrophic performance degradation, which we have observed on several models.

**Comment by suharshs** [comment]
Thanks ChanZhou, I was able to run your code. 

Thanks Feng, yes the value of the output tensor is off b y 0.115.. which is exactly the value of the scale. So the output is off by 1 LSB for some values.

Interestingly, I was able to reproduce this error when running a floating point model (Which has the FakeQuant operations in the TFLite graph), but the error occurred in fewer values.


    [[ 0 72 96  3]]
    Max absolute error: 0.11157035827636719. Example values tf: 1.561981201171875; tflite: 
    1.4504108428955078

This makes me believe that there is a discrepancy between the constant folding of FakeQuant in TensorFlow and the constant folding of FakeQuant in TensorFlow Lite, that results in a mismatch before quantization even happens.

Will verify that theory and get back.

**Comment by liufengdb** [comment]
ChanZou, I checked the value range of the bias and also the output range, it is [-11.9, 16.5]. I would suggest applying Relu6 or on the output to reduce this range. It should help the accuracy for deep models.

**Comment by ChanZou** [comment]
Feng, thanks for the suggestion. This is a minimal example helping us to understand what shall we expect from TFLite down to the 'atomic' level. I understand narrower range for the activation will help with the precision of both bias and the activation itself. Yet we are still seeing quite a significant drop in regression type of tasks.

**Comment by suharshs** [comment]
I have verified that the different in the floating point graphs is due to the converter changing FakeQuant to Quantize and Dequantize. Resulting in numerical difference.

That being said, this is unrelated to your original issue, which after looking into I think is fundamental differences in rounding / rescaling in FakeQuant vs the actual tflite kernels. 

It seems that this particular model is very prone to these differences, the most practical approach to resolving some of these issues would be to explore the following model-level changes (from simplest to try to more difficult):

1) Try post-training quantization which uses per-channel quantization which may resolve some issues.
2) Try using a model with restricted activation ranges (as Feng suggested). I have personally seen this work both ways, but since your model has a such a large range for activations this may be very helpful.
3) Maybe try stochastic rounding in the FakeQuant operation. (this isn't in the code)

To help with suggestions, couple questions:

1) Without this rounding error from FakeQuant to real quant are you getting sufficient accuracy in you FakeQuant float TF model?
2) How far are you from your target accuracy for your problem with FakeQuant?

**Comment by ChanZou** [comment]
To answer your questions:
  1. Yes, we are getting very close performance when comparing fake-quantized TF model with the vanilla one (without fake-quantization nodes) so it is particularly surprising to see such a discrepancy at the 'final' step.
  2. We are seeing up to ~30% performance drop in some tasks. Can you provide some insights on how often do you/your team see this scale of discrepancy?

Thank you for your suggestions, and a couple of questions/comments on them if you don't mind:
  1. Is there a more specific reason why per-channel quantization help other than potentially narrower range? Does it share the same kernel with the per-layer quantization and hence suffer from the same issue?
  2. As this is just a toy example, we have activations attached to most Conv2D, and still seeing errors accumulate into quite significant discrepancies across layers.

Again thank you for spending time on this.

**Comment by suharshs** [comment]
Couple more things to confirm.

Can you confirm that your trained float model without FakeQuant can be converted to TFLite and doesn't have the 30% drop? Just for sanity check. :)

**Comment by AmirKhooj** [comment]
Hey @ChanZou, 

What is the status of this issue? I also am seeing quite big amount of discrepancy between a quantized model and its TFLITE converted version. My environment is similar to your except that I am using Tensorflow 1.15.2 and Python 3.6. I use the `tflite_convert` command line tool. My input float model is a quantized aware trained model (hence with FakeQuantize nodes).  The command line I run is this:

```
tflite_convert \
  --graph_def_file=my_frozen_graph.pb \
  --output_file=my_tflite_model.tflite \
  --output_format=TFLITE \
  --input_shape=1,257,257,3 \
  --input_arrays="MobilenetV2/MobilenetV2/input" \
  --output_arrays="MobilenetV2/expanded_conv/input" \
  --inference_type=QUANTIZED_UINT8\
  --inference_input_type=QUANTIZED_UINT8 \
  --std_dev_values=128 \
  --mean_values=128 \
  --change_concat_input_ranges=true
```

Similar to what you mentioned above, the TFLITE model has the quantization statistics for converting the quantized values to float (scale and bias). When I use this statistic to compare with the float model, there is a big difference, I am seeing maximum of 0.2 difference and this is the output of only one convolution block.

**Comment by ChanZou** [comment]
Hi @AmKhG 
In our case, the discrepancy for individual images remains 'arbitrary', that is, it is hard to conclude an upper or lower bound for the discrepancy. However, if looking at the aggregated stats (mAP) for the test set as a whole, the performance will not be having a siginificant drop. Intuitively this might have something to do with the ergodicity. Hope this help.

**Comment by sushreebarsa** [comment]
@ChanZou  Could you please try on latest stable version of tf 2.5 or 2.4.1 and let us know if this is still an issue.Thanks!
